### PR TITLE
Rework wallet objects hierarchy

### PR DESCRIPTION
**Pull request by abi87**
Removed need for slicing CWalletTx/Cert. Introduced UTs for wallet

### CODE REVIEW

File: src/wallet/gtest/test_wallet_cert.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+class CertInWalletTest : public ::testing::Test {
+public:
+    CertInWalletTest(): walletDbLocation(), pWallet(nullptr), pWalletDb(nullptr) {}
+    ~CertInWalletTest() = default;
+
+    void SetUp() override {
+        //Setup environment
+        SelectParams(CBaseChainParams::TESTNET);
+        boost::filesystem::path walletDbLocation = boost::filesystem::temp_directory_path() / boost::filesystem::unique_path();
```
Comment (alsala) [inline src/wallet/gtest/test_wallet_cert.cpp:18]: use a different name, just to disambiguate with data member and improve readability

File: src/wallet/gtest/test_wallet_cert.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        pWallet = new CWallet("wallet.dat");
+        try {  pWalletDb = new CWalletDB(pWallet->strWalletFile, "cr+"); }
+        catch(std::exception& e) {
+            ASSERT_TRUE(false)<<"Could not create tmp wallet db for reason "<<e.what();
+        }
+    };
+
+    void TearDown() override {
+        std::vector<std::shared_ptr<CWalletTransactionBase> > vWtx;
+        DBErrors nZapWalletRet = pWallet->ZapWalletTx(vWtx);
```
Comment (alsala) [inline src/wallet/gtest/test_wallet_cert.cpp:32]: nZapWalletRet is not used, should it be checked?

File: src/zcbenchmarks.cpp [hunk 3]
```diff
@@ -291,11 +291,11 @@ double benchmark_try_decrypt_notes(size_t nAddrs)
     }
 
     auto sk = libzcash::SpendingKey::random();
-    auto tx = GetValidReceive(*pzcashParams, sk, 10, true);
+    auto walletTx = GetValidReceive(*pzcashParams, sk, 10, true);
 
     struct timeval tv_start;
     timer_start(tv_start);
-    auto nd = wallet.FindMyNotes(tx);
+    auto nd = wallet.FindMyNotes(walletTx.getWrappedTx());
```
Comment (alsala) [inline src/zcbenchmarks.cpp:298]: nd does not seem to be used? (It was the same in the original code anyway)

File: src/wallet/gtest/test_wallet_cert.cpp (hunk 2 above)
Comment (abi87) [reply src/wallet/gtest/test_wallet_cert.cpp:32]: the `ZapWalletTx `here makes sure that the wallet is duly cleaned-up and different UTs are well decoupled.
I would check `ZapWalletTx `return type with an EXPECT_TRUE(), in order to signal potential coupling of UTs, should `ZapWalletTx `fail.

File: src/wallet/gtest/test_wallet_cert.cpp (hunk 1 above)
Comment (abi87) [reply src/wallet/gtest/test_wallet_cert.cpp:18]: Actually I intended yo use the data member, removing declaration here

File: src/zcbenchmarks.cpp (hunk 3 above)
Comment (abi87) [reply src/zcbenchmarks.cpp:298]: It's a benchmark, so I believe they want to verify the time it takes for finding notes

**Pull request merged**